File: conformance/reference-js/static-data.js
'use strict';

function failure(code) {
  return { ok: false, code };
}

function resolveStaticValue(input) {
  if (input.data.source === 'static') return { ok: true, value: input.data.value };
  if (input.data.source !== 'ref') return failure('STATIC_DATA_REF_INVALID');
  const target = input.datasources?.[input.data.ref];
  if (!target || target.source !== 'static') return failure('STATIC_DATA_REF_INVALID');
  if (input.data.responseMapping !== undefined || target.responseMapping !== undefined) {
    return failure('STATIC_RESPONSE_MAPPING_NOT_ALLOWED');
  }
  return { ok: true, value: target.value };
}

function resolveStaticData(input) {
  if (input.data.responseMapping !== undefined) return failure('STATIC_RESPONSE_MAPPING_NOT_ALLOWED');
  const resolved = resolveStaticValue(input);
  if (!resolved.ok) return resolved;
  const value = resolved.value;
  if ((input.component === 'table' || input.component === 'chart') && !Array.isArray(value)) {
    return failure('STATIC_DATA_SHAPE_MISMATCH');
  }
  if (input.component === 'statCard' || input.component === 'text') {
    const scalarOrObject = value !== null && !Array.isArray(value)
      && (typeof value !== 'object' || Object.getPrototypeOf(value) === Object.prototype);
    if (!scalarOrObject) return failure('STATIC_DATA_SHAPE_MISMATCH');
    if (input.props?.valueField) {
      if (typeof value !== 'object'
        || !Object.prototype.hasOwnProperty.call(value, input.props.valueField)) {
        return failure('STATIC_DATA_SHAPE_MISMATCH');
      }
      return { ok: true, value: value[input.props.valueField], network: false };
    }
  }
  return { ok: true, value, network: false };
}

module.exports = { resolveStaticData };

File: conformance/reference-python/static_data.py
def failure(code):
    return {"ok": False, "code": code}


def resolve_static_value(input_value):
    data = input_value["data"]
    if data["source"] == "static":
        return {"ok": True, "value": data["value"]}
    if data["source"] != "ref":
        return failure("STATIC_DATA_REF_INVALID")
    target = input_value.get("datasources", {}).get(data.get("ref"))
    if not isinstance(target, dict) or target.get("source") != "static":
        return failure("STATIC_DATA_REF_INVALID")
    if "responseMapping" in data or "responseMapping" in target:
        return failure("STATIC_RESPONSE_MAPPING_NOT_ALLOWED")
    return {"ok": True, "value": target.get("value")}


def resolve_static_data(input_value):
    if "responseMapping" in input_value["data"]:
        return failure("STATIC_RESPONSE_MAPPING_NOT_ALLOWED")
    resolved = resolve_static_value(input_value)
    if not resolved["ok"]:
        return resolved
    value = resolved["value"]
    component = input_value["component"]
    if component in ("table", "chart") and not isinstance(value, list):
        return failure("STATIC_DATA_SHAPE_MISMATCH")
    value_field = input_value.get("props", {}).get("valueField")
    if component in ("statCard", "text"):
        if value is None or isinstance(value, list) or not isinstance(value, (dict, str, int, float, bool)):
            return failure("STATIC_DATA_SHAPE_MISMATCH")
        if value_field:
            if not isinstance(value, dict) or value_field not in value:
                return failure("STATIC_DATA_SHAPE_MISMATCH")
            return {"ok": True, "value": value[value_field], "network": False}
    return {"ok": True, "value": value, "network": False}
